lib/ide/src/controller.rs

```diff
@@ -0,0 +1,426 @@
+//! This module contains all the controllers. They cover everything that is
+//! between clients of remote services (like language server and file manager)
+//! and views.
+//!
+//! The controllers create a tree-like structure, with project controller being
+//! a root, then module controllers below, then graph/text controller and so on.
+//!
+//! As a general rule, while the "upper" (i.e. closer to root) nodes may keep
+//! handles to the "lower" nodes (e.g. to allow their reuse), they should never
+//! manage their lifetime.
+//!
+//! Primarily views are considered owners of their respective controllers.
+//! Additionally, controllers are allowed to keep strong handle "upwards".
+//!
+//! Controllers store their handles using `utils::cell` handle types to ensure
+//! that mutable state is safely accessed.
+
+use crate::prelude::*;
+
+/// General-purpose `Result` supporting any `Error`-compatible failures.
+pub type FallibleResult<T> = Result<T,failure::Error>;
+
+/// Macro defines `StrongHandle` and `WeakHandle` newtypes for handles storing
+/// the type given in the argument.
+///
+/// This allows treating handles as separate types and fitting them with impl
+/// methods of their own. While not necessary, such implementation may allow
+/// hiding from user gritty details of `with` usage behind nice, easy API.
```

Comment:
Im reading this file from top to bottom, and I see again a word "handle" but without explanation - what is this handle and what it is used for? I mean - if that was explained in the module comment, I would not ask about it.